### DISCUSSION TITLE
Configurable file storage adapter per-field [with implementation]

### DISCUSSION THREAD

**Discussion by JoshTheDerf** [directus/directus#14436] [Feature Requests]
## Description
There's been a bit of discussion on providing configurable storage adapters accessible from the admin app, a feature that I happen to have use for.
- [#8878](https://github.com/directus/directus/discussions/8878) mentions a potential workaround.
- [#11148](https://github.com/directus/directus/discussions/11148) proposes configurable storage adapters per-folder, but no resolution was reached on the best way to implement this without imposing arbitrary restrictions on the data model.

Since the API already supports swapping storage adapters during the upload process, and we already have a way of configuring target folders per-field, it seems straightforward to extend field options to also allow selecting a storage adapter. This wouldn't impose any new limitations on the data model (I think) and allows the existing behavior (choose the default storage adapter) to remain as is.

This seems like a straightforward change, so I've gone ahead and made an initial functional implementation at [JoshTheDerf/directus/tree/feature/configure-interface-file-storage](https://github.com/directus/directus/compare/main...JoshTheDerf:directus:feature/configure-interface-file-storage)

Implementation notes:
- Exposes an additional property on serverStore with storage names and drivers. (Most other storage config information is stripped out for security.) Wasn't sure how to get a list of storage options to the client, there's likely a more "correct" way to do this.
   ![image](https://user-images.githubusercontent.com/5861371/179071137-ced6904d-803a-479f-8213-212808b610a3.png)
- Adds a new system interface: `system-storage`. 

Screenshot:
![image](https://user-images.githubusercontent.com/5861371/179070267-19cc4cde-e2a5-40f3-864f-2eef236741bf.png)

What are your thoughts on this potential feature?

Thanks a heap for all y'alls hard work on Directus! Seems like it's come a long way in recent years, and I'm finding the code quite enjoyable to hack on. 😄 

## Type of Change

- [ ] Bugfix
- [ ] Improvement
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [ ] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated

**Comment by rijkvanzanten** [comment]
Would Storage Provider also filter the "select existing" list to that particular storage adapter? 

I think it's a great feature to have _but_ we should handle it as one big feature to allow the storage selection and usage everywhere it's applicable, rather than only as a single interface option for a single interface. For example, when uploading a file directly to the file library, we should also expose the selection of storage adapter. When we go that route, we should also add a way where people can _change_ the storage adapter of a file, and the file should be moved to the corresponding adapter automatically 🤔 

I think this is one of those features that might balloon out of "scope" very quickly 🎈 🙂

**Reply by JoshTheDerf** [reply]
Agreed, filtering the select existing list makes sense since the feature is modeled after the folder selection feature which does filter the folder for "select existing".

Adding the selection for uploading a folder to the library also seems like a good idea, and would hopefully be straightforward to implement with the `system-storage` interface in place, just need to figure out how to work it into the file upload dialog, since it starts uploading immediately upon file selection. (And how should it work with drag-and-drop?)

I don't think changing the storage provider needs to be in-scope here, in the same way that changing the selected folder only applies to newly uploaded files.

**Reply by benhaynes** [reply]
Just to add on here, I am a big fan of this feature request. I think we might want to approach this in a way that keeps the files bucketed within the storage adapters. So on the File Library, the root nav on the left is adapters, then folders within. Perhaps a default is set (first in the ENV?) that dictates where basic drag and drop files go, and you can use the modal upload to choose specifically.

Therefore, changing a file's adapter would be uploading new and then deleting the original.

For interfaces, I think we allow setting a storage adapter, default folder within (which we have), and discuss having a toggle to restrict the field to that one adapter (or not).

There might be a way around this scope ballooning, with smart defaults and K.I.S.S.

**Reply by JoshTheDerf** [reply]
Sounds like the idea of tying storage selection to the folder collection, as proposed here?
https://github.com/directus/directus/discussions/11148#discussioncomment-2424822

I definitely like that idea, since it makes it transparently obvious where each folder actually stores its contents and requires no additional selection of storage adapters during the file upload process.

Migration would be tricky though, as there would need to be legacy support for folders with files in a variety of storage adapters to account for the existing setup. And as mentioned in the thread linked above, it reduces the flexibility of directus' file storage system a bit, for better or worse.

I'd be happy with any of the proposed solutions as long as there was some way to put data for different fields in different storage locations, be it defined by folder or directly.

**Reply by benhaynes** [reply]
> Sounds like the idea of tying storage selection to the folder collection, as proposed here?

Hmm, well, I believe that is talking about _syncing_ folders on the storage to our virtual folders. I'm just proposing that we use storage adapters as the root in the File Library instead of commingling them together within the virtual folders. (Hope that makes sense!) But similar benefits: you know where the files/folders are located.

**Reply by JoshTheDerf** [reply]
If I understand you correctly, could that be implemented with minimal data model changes by having a configurable storage adapter field on any top-level folders, then forcing all folders and files under them to inherit that storage adapter?

That would allow a logical grouping of root-level folders by storage adapter and could be *presented* as storage adapters being at the root of the folder tree.

Then if the root level storage adapter changes, propagate storage adapter configuration down through the tree, moving files between old and new adapter as needed?

**Reply by benhaynes** [reply]
Haha, I defer to @rijkvanzanten on the actual implementation. But at the surface, I imagine something like this... though, not sure if the folders should be within the adapters or separate like in this mockup.

![File Library](https://user-images.githubusercontent.com/522079/179317157-6e10fa00-6ee6-4cb8-ab0f-9f8d7089a744.png)

**Reply by rijkvanzanten** [reply]
For what it's worth, our current virtual folders aren't linked in any way to the storage adapter, so I think we'd consider the storage adapter links in the sidebar more as filter shortcuts

**Reply by benhaynes** [reply]
Right, that's why I have the mockup like that. 👍

**Reply by JoshTheDerf** [reply]
@rijkvanzanten Would it be worth it for me to take a stab at implementing something along the lines of that mockup or does the idea need to be made more concrete first?

**Reply by rijkvanzanten** [reply]
We'll have to think through whether we want folders to stay "global" / "virtual", or if each storage adapter would become a new "root file library" 🤔 I think I would personally expect each storage adapter to have it's own files (I kinda see each storage adapter as another USB stick you plug in, each with their own root folder structure)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
